llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp

```diff
@@ -1073,6 +1073,16 @@ Instruction *InstCombinerImpl::visitFMul(BinaryOperator &I) {
     return Result;
   }
 
+  // tan(X) * cos(X) -> sin(X)
+  if (I.hasAllowReassoc() && Op0->hasOneUse() && Op1->hasOneUse()) {
+    Value *X;
+    if (match(Op0, m_Intrinsic<Intrinsic::tan>(m_Value(X))) &&
+        match(Op1, m_Intrinsic<Intrinsic::cos>(m_Specific(X)))) {
+      Value *Sin = Builder.CreateUnaryIntrinsic(Intrinsic::sin, X, &I);
```

Comment:
@arsenm Is `sin/sinf` always available if `tan/cos` are supported?

Reply:
I think in practice this will be true. This system is a mess though. In principle the intrinsics should always work regardless of target libcalls. We gate transforms on pure intrinsics based on libcall legality, but it's all hacks on hacks .